### PR TITLE
Perform action

### PR DESCRIPTION
**Pull request by aaronchongth**
This adds support for `perform_action`, targeting specific mission names that have been predefined in the MiR dashboard, and can be listed in the config file to be called as a `perform_action` task.

### CODE REVIEW

**Review by mxgrey — CHANGES_REQUESTED** [review]
I left a few comments about things that I think should be changed, but also some suggestions that are just meant to be notes for potential future improvements.

File: configs/mir_config.yaml [hunk 1]
```diff
@@ -93,3 +96,4 @@ rmf_fleet:
     delivery: False
     clean: False
     finishing_request: "nothing" # [park, charge, nothing]
+    action_categories: ["dock_to_charger", "custom_mission_1"]
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply configs/mir_config.yaml:99]: Yup that definitely sounds like a much better way to handle `perform_action`, thanks! I've created the issue, https://github.com/osrf/fleet_adapter_mir/issues/19

File: fleet_adapter_mir/MiRCommandHandle.py [hunk 2]
```diff
@@ -543,6 +555,62 @@ def estimate_arrival(self, goal, current_position):
...
+            # Load self.action_check_task_completion
+            def _check_task_completion():
+                mission_status = self.mir_api.mission_queue_id_get(mission_queue_id)
+                if 'finished' in mission_status and mission_status['finished'] is not None:
+                    return True
+                return False
+            self.action_check_task_completion = _check_task_completion
+
+            # Keep track of perform action
+            self.node.get_logger().warn(f"Robot [{self.name}] starts [{category}] action")
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply fleet_adapter_mir/MiRCommandHandle.py]: https://github.com/osrf/fleet_adapter_mir/pull/15/commits/1e412a7f292e350aaeeaa607c1cba44bedef4521

File: fleet_adapter_mir/MiRCommandHandle.py [hunk 3]
```diff
@@ -543,6 +555,62 @@ def estimate_arrival(self, goal, current_position):
...
+                         execution: adpt.robot_update_handle.ActionExecution):
+        with self._update_mutex:
+            # Get list of missions, get guid of mission
+            missions = self.mir_api.missions_get()
+            mission_guid = None
+            for m in missions:
+                if m['name'] == category:
+                    mission_guid = m['guid']
+            if mission_guid is None:
+                self.node.get_logger().error(f'Action category {category} not supported, ignoring')
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply fleet_adapter_mir/MiRCommandHandle.py]: https://github.com/osrf/fleet_adapter_mir/pull/15/commits/6dcefd53abaa53d440b0d0637816449ad4ae24a7

File: fleet_adapter_mir/MiRCommandHandle.py [hunk 4]
```diff
@@ -543,6 +555,62 @@ def estimate_arrival(self, goal, current_position):
...
+                    f"action [{self.action_category}] is completed")
+                self.action_execution.finished()
+            else:
+                self.node.get_logger().warn(
+                    f"action [{self.action_category}] is killed/canceled")
+            self.action_execution = None
+            self.action_category = None
+            self.action_description = None
+            self.action_start_time = None
+            self.action_check_task_completion
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply fleet_adapter_mir/MiRCommandHandle.py]: https://github.com/osrf/fleet_adapter_mir/pull/15/commits/24707585d1560293d96b8084df7798c579f20daf

File: fleet_adapter_mir/MiRCommandHandle.py [hunk 5]
```diff
@@ -543,6 +555,62 @@ def estimate_arrival(self, goal, current_position):
...
+                    return True
+                return False
+            self.action_check_task_completion = _check_task_completion
+
+            # Keep track of perform action
+            self.node.get_logger().warn(f"Robot [{self.name}] starts [{category}] action")
+            self.action_category = category
+            self.action_description = description
+            self.action_start_time = self.node.get_clock().now()
+            self.action_execution = execution
```
[the remark this answers is not in the document]
Comment (aaronchongth) [reply fleet_adapter_mir/MiRCommandHandle.py]: Thanks for the suggestion, it is definitely much cleaner now.
https://github.com/osrf/fleet_adapter_mir/pull/15/commits/80e30203af3b7961012c7cce033f28a2a29d8c78

**Review by mxgrey — APPROVED** [review]
Thanks for the changes!

**Pull request merged**